### PR TITLE
vim-patch:8.1.1564, 1570, 1623, 1712 : sign column takes up space

### LINKED ISSUE #4515: Allow to show signcolumn on top of numbers column
The current available options for `signcolumn` are `auto`, `yes`, and `no`, all of whom have their annoyances:

- `no` - doesn't allow to highlight lines at all
- `yes` - always takes space in vim even when no errors or warning as highlighted
- `auto` - makes text in vim "jump" when error or warning happens

**Describe the solution you'd like**

An extra option `number` that makes markers to be visible *on top* of numbers column. This solution has following advantages:

- no space is taken when there are no errors or warnings
- text is not jumping when errors or warnings start to be highlighted

### PR DESCRIPTION
**Pull request by Shougo**
Problem:    Sign column takes up space.  (Adam Stankiewicz)
Solution:   Optionally put signs in the number column. (Yegappan Lakshmanan,
            closes vim/vim#4555, closes vim/vim#4515)
https://github.com/vim/vim/commit/394c5d8870b15150fc91a4c058dc571fd5eaa97e

~~I have tried the porting.  But it does not work as well...~~

Any advice and improvements are is appreciated!

### CODE REVIEW

**Comment by Shougo** [comment]
Hm.  It has SEGV when `signcolumn=number` is enabled.

**Comment by erw7** [comment]
> Any advice and improvements are is appreciated!

@Shougo I think the following changes need to be made.

```C
diff --git a/src/nvim/screen.c b/src/nvim/screen.c
index bbfa16b33..17384b033 100644
--- a/src/nvim/screen.c
+++ b/src/nvim/screen.c
@@ -2856,7 +2856,7 @@ win_line (
           if (count > 0) {
               get_sign_display_info(false, wp, lnum, row,
                       startrow, filler_lines, filler_todo, count,
-                      &c_extra, &c_final, extra, &p_extra, &n_extra,
+                      &c_extra, &c_final, extra, sizeof(extra), &p_extra, &n_extra,
                       &char_attr, &draw_state, &sign_idx);
           }
       }
@@ -2876,7 +2876,7 @@ win_line (
             int count = win_signcol_count(wp);
             get_sign_display_info(true, wp, lnum, row,
                 startrow, filler_lines, filler_todo, count,
-                &c_extra, &c_final, extra, &p_extra, &n_extra,
+                &c_extra, &c_final, extra, sizeof(extra), &p_extra, &n_extra,
                 &char_attr, &draw_state, &sign_idx);
           } else if (row == startrow + filler_lines) {
             // Draw the line number (empty space after wrapping). */
@@ -4479,6 +4479,7 @@ static void get_sign_display_info(
     int *c_extrap,
     int *c_finalp,
     char_u *extra,
+    size_t extra_size,
     char_u **pp_extra,
     int *n_extrap,
     int *char_attrp,
@@ -4508,7 +4509,7 @@ static void get_sign_display_info(
         *c_finalp = NUL;
 
         if (nrcol) {
-          sprintf((char *)extra, "%-*s ", number_width(wp),
+          snprintf((char *)extra, extra_size, "%-*s ", number_width(wp),
               *pp_extra);
           *pp_extra = extra;
           *n_extrap = (int)STRLEN(*pp_extra);
@@ -4523,12 +4524,12 @@ static void get_sign_display_info(
           *n_extrap = symbol_blen +
             (win_signcol_width(wp) - mb_string2cells(*pp_extra));
 
-          assert(sizeof(extra) > (size_t)symbol_blen);
-          memset(extra, ' ', sizeof(*extra));
+          assert(extra_size > (size_t)symbol_blen);
+          memset(extra, ' ', extra_size);
           memcpy(extra, *pp_extra, symbol_blen);
 
           *pp_extra = extra;
-          *pp_extra[*n_extrap] = NUL;
+          (*pp_extra)[*n_extrap] = NUL;
         }
       }
       *char_attrp = sign_get_attr(text_sign, SIGN_TEXT);
```

**Comment by Shougo** [comment]
@erw7 Thank you!  SEGV is fixed now.
But the tests are failed.

```
From test_signs.vim:
Executed Test_sign() in   0.000297 seconds
Executed Test_sign_change_type() in   0.000182 seconds
Executed Test_sign_completion() in   0.000260 seconds
Executed Test_sign_cursor_position() in   0.000331 seconds
    Skipped
Executed Test_sign_delete_buffer() in   0.000183 seconds
Executed Test_sign_funcs() in   0.000293 seconds
Executed Test_sign_group() in   0.000305 seconds
Executed Test_sign_id_autogen() in   0.000271 seconds
Executed Test_sign_invalid_commands() in   0.000277 seconds
Executed Test_sign_jump_func() in   0.000157 seconds
Executed Test_sign_lnum_adjust() in   0.000266 seconds
Executed Test_sign_memfailures() in   0.000260 seconds
    Skipped
Executed Test_sign_numcol() in   0.000164 seconds
Executed Test_sign_priority() in   0.000291 seconds
Executed Test_sign_undefine_still_placed() in   0.000267 seconds
Executed Test_sign_unplace() in   0.000177 seconds
Executed 16 tests in   0.177111 seconds
4 FAILED:
Found errors in Test_sign():
Caught exception in Test_sign(): Vim(new):E325: ATTENTION @ function RunTheTest[44]..Test_sign, line 42
Found errors in Test_sign_group():
Caught exception in Test_sign_group(): Vim(split):E325: ATTENTION @ function RunTheTest[44]..Test_sign_group, line 207
Found errors in Test_sign_jump_func():
Caught exception in Test_sign_jump_func(): Vim(edit):E325: ATTENTION @ function RunTheTest[44]..Test_sign_jump_func, line 5
Found errors in Test_sign_numcol():
function RunTheTest[44]..Test_sign_numcol line 11: Expected '=> 01234' but got '=>01234 '
function RunTheTest[44]..Test_sign_numcol line 24: Expected '=>01234' but got '01234  '
SKIPPED Test_sign_cursor_position(): cannot make screendumps
SKIPPED Test_sign_memfailures(): Nvim does not support test_alloc_fail()
```

I need to investigate the reason...

**Comment by Shougo** [comment]
I cannot fix the last errors...

```
Found errors in Test_sign_numcol():
function RunTheTest[44]..Test_sign_numcol line 11: Expected '=> 01234' but got '=>01234 '
function RunTheTest[44]..Test_sign_numcol line 24: Expected '=>01234' but got '01234  '
SKIPPED Test_sign_cursor_position(): cannot make screendumps
SKIPPED Test_sign_memfailures(): Nvim does not support test_alloc_fail()
```

File: src/nvim/screen.c
Comment (erw7) [inline src/nvim/screen.c:2885]: > I cannot fix the last errors...

I think the problem is that you are making this place an `else if` statement. You must include up to line 2940 in the `else` statement.

Comment (Shougo) [reply src/nvim/screen.c]: Thanks.

**Comment by Shougo** [comment]
```
Found errors in Test_sign_numcol():
function RunTheTest[44]..Test_sign_numcol line 24: Expected '=>01234' but got '01234  '
SKIPPED Test_sign_cursor_position(): cannot make screendumps
SKIPPED Test_sign_memfailures(): Nvim does not support test_alloc_fail()
```

The only one error exists...

**Comment by Shougo** [comment]
Fixed lint errors.

File: src/nvim/screen.c
Comment (erw7) [inline src/nvim/screen.c:4523]: You have a new enbug here when fixing the lint error.

```C 
snprintf((char *)extra, extra_size, "%-*s ",
              number_width(wp), *pp_extra);
```

**Comment by erw7** [comment]
> The only one error exists...

The following changes will make all the tests(`src/nvim/tesdir/test_signs.vim`, `test/functional/ui/sign_spec.lua`) pass.

```diff
diff --git a/src/nvim/option.c b/src/nvim/option.c
index 2564f1710..3fe7beefb 100644
--- a/src/nvim/option.c
+++ b/src/nvim/option.c
@@ -7397,7 +7397,8 @@ int win_signcol_count(win_T *wp)
   int maximum = 1, needed_signcols;
   const char *scl = (const char *)wp->w_p_scl;

-  if (*scl == 'n') {
+  if (*scl == 'n' && (*(scl + 1) == 'o'
+                      || (*(scl + 1) == 'u' && (wp->w_p_nu || wp->w_p_rnu)))) {
     return 0;
   }
   needed_signcols = buf_signcols(wp->w_buffer);
diff --git a/src/nvim/screen.c b/src/nvim/screen.c
index 70df2fd36..e4bf98957 100644
--- a/src/nvim/screen.c
+++ b/src/nvim/screen.c
@@ -2855,7 +2855,7 @@ win_line (
           /* Show the sign column when there are any signs in this
            * buffer or when using Netbeans. */
           int count = win_signcol_count(wp);
-          if (count > 0 && !is_sign_number) {
+          if (count > 0) {
               get_sign_display_info(
                   false, wp, lnum, row,
                   startrow, filler_lines, filler_todo, count,
```

File: src/nvim/screen.c
Comment (Shougo) [reply src/nvim/screen.c:4523]: Thanks.

**Comment by Shougo** [comment]
Thank you!  The tests are passed!

**Comment by sentriz** [comment]
nice! thank you Shougo and erw7

**Comment by Shougo** [comment]
The failed reason is...

```
From test_textformat.vim:

Executed Test_format_align() in   0.001552 seconds

Executed Test_format_list_auto() in   0.001123 seconds

Executed Test_format_para() in  cat: write error: Resource temporarily unavailable

Makefile:181: recipe for target 'newtests' failed

make[1]: *** [newtests] Error 1

make[1]: Leaving directory '/home/travis/build/neovim/neovim/src/nvim/testdir'

Makefile:122: recipe for target 'oldtest' failed

make: *** [oldtest] Error 2
```

Too bad.

File: src/nvim/option.c
Comment (erw7) [inline src/nvim/option.c:7407]: You might want to add a comment.

File: src/nvim/screen.c
Comment (erw7) [inline src/nvim/screen.c:2819]: It is an unnecessary blank line.

File: src/nvim/screen.c
Comment (erw7) [inline src/nvim/screen.c:4480]: I think `bool` is better.

File: src/nvim/screen.c
Comment (erw7) [inline src/nvim/screen.c:7580]: It is an unnecessary blank line.

File: src/nvim/testdir/test_signs.vim
Comment (erw7) [inline src/nvim/testdir/test_signs.vim:9]: It's not present in the original vim patch, is this necessary? In my environment, the test passes without it.

File: src/nvim/testdir/test_signs.vim
Comment (erw7) [inline src/nvim/testdir/test_signs.vim:1788]: For some reason there seems to be one too many space.

https://ci.appveyor.com/project/neovim/neovim/builds/34075588/job/ao12eot4yxwssrip#L13543
```
From test_signs.vim:
Found errors in Test_sign_numcol():
function RunTheTest[44]..Test_sign_numcol line 30: Expected '=>  01234' but got ' => 01234'
../../../build/bin/nvim.exe -u NONE -U NONE -i NONE --noplugin --headless -S summarize.vim messages
Test results:
```

Comment (erw7) [reply src/nvim/testdir/test_signs.vim:1788]: It seems that https://github.com/vim/vim/commit/4dff4aed09d2b0d570ca0d19de9cb08bdf03e695 is missing. I don't think this is necessary, but it seems to work differently with or without it compared to vim.

File: src/nvim/testdir/test_signs.vim
Comment (Shougo) [reply src/nvim/testdir/test_signs.vim:9]: I don't pass... Because

```
Found errors in Test_sign():
Caught exception in Test_sign(): Vim(new):E325: ATTENTION @ function RunTheTest[44]..Test_sign, line 42
Found errors in Test_sign_group():
Caught exception in Test_sign_group(): Vim(split):E325: ATTENTION @ function RunTheTest[44]..Test_sign_group, line 207
Found errors in Test_sign_jump_func():
Caught exception in Test_sign_jump_func(): Vim(edit):E325: ATTENTION @ function RunTheTest[44]..Test_sign_jump_func, line 5
```

I think swapfile is on in neovim default.

File: src/nvim/testdir/test_signs.vim
Comment (Shougo) [reply src/nvim/testdir/test_signs.vim]: OK.  I have changed the test code.

**Comment by Shougo** [comment]
Finally, all tests are passed!

File: src/nvim/testdir/test_signs.vim
Comment (erw7) [reply src/nvim/testdir/test_signs.vim:10]: > I don't pass... 

I think that's because the swap file is still there when the test ended abnormally by `SEGV`. Therefore, instead of `setting noswapfile` here, you should delete any remaining swap files in your environment.

Comment (Shougo) [reply src/nvim/testdir/test_signs.vim]: OK.  I have removed it.

**Comment by Shougo** [comment]
I have resolved all comments and all tests are passed.
Can you merge it?

**Comment by antoinemadec** [comment]
@erw7 , @marvim , is there anything blocking this merge?

**Comment by teto** [comment]
@antoinemadec it would be nice to have your feedback too, did it work for you ? how long have you tested it ? Code reviews are nice but it's helpful to have user feedback too.

**Comment by sentriz** [comment]
just did some testing. is working well for me :)

another thing though, `:echo has("patch-8.1.1564")` returns `0`. should it be 1?

**Comment by fsouza** [comment]
@Shougo can you rebase this branch? If so, I can use for the next couple of days to confirm it works 😁 

If not, I can create a custom branch on my fork too, no worries.

**Comment by jamessan** [comment]
> another thing though, :echo has("patch-8.1.1564") returns 0. should it be 1?

Not until after we finish merging the 8.0 patches.

**Comment by antoinemadec** [comment]
> @antoinemadec it would be nice to have your feedback too, did it work for you ? how long have you tested it ? Code reviews are nice but it's helpful to have user feedback too.

@teto , I did today. Here are the main things I tried:

- setup: a lot of plugins (including coc.nvim).
- create git diffs and LSP errors on the same line: each creating signs
- use multiple tabs, cycle though them fast
- resizing my terminal

It seemed to work just fine for me :smiley: 

Cheers,
Antoine

**Comment by Shougo** [comment]
> @Shougo can you rebase this branch? If so, I can use for the next couple of days to confirm it works grin

I have rebased the branch now.

**Comment by fsouza** [comment]
@Shougo thanks! I created a branch earlier today and everything's been working fine here :D

**Comment by fsouza** [comment]
Just confirming: been using it over the last 3 days with the builtin lsp client and it works like a charm! 😁

**Comment by antoinemadec** [comment]
@jamessan @teto @Shougo @erw7 , are there other PRs blocking before we can merge this one?
The hype is strong with this one :smile:

**Review by erw7 — APPROVED** [review]

**Comment by Shougo** [comment]
Thank you for the approve.

**Comment by teto** [comment]
vim implementation is a bit ugly IMO and the result could be improved too
2020-07-25-000244_601x316_scrot
It's easier to merge though until we get a better solution.

**Comment by Shougo** [comment]
Thank you for the merge!

**Comment by Shougo** [comment]
> vim implementation is a bit ugly IMO and the result could be improved too

We need more active developers...

**Pull request merged**
